backend/src/main/java/com/bakdata/conquery/apiv1/QueryProcessor.java

```diff
@@ -19,20 +29,38 @@ public QueryProcessor(Namespaces namespaces, ConqueryConfig config) {
 		this.config = config;
 	}
 
-	public SQStatus postQuery(Dataset dataset, IQuery query, URLBuilder urlb) throws JSONException {
+	public SQStatus postQuery(Dataset dataset, IQuery query, URLBuilder urlb, User user) throws JSONException {
+		authorize(user, dataset.getId(), Ability.READ);
 		Namespace namespace = namespaces.get(dataset.getId());
+		
+		// Check dataset
+		user.checkPermission(new DatasetPermission(null, EnumSet.of(Ability.READ), dataset.getId()));
+		
+		// Check reused query
+		for(ManagedQueryId requiredQueryId : query.collectRequiredQueries()) {
+			user.checkPermission(new QueryPermission(null, EnumSet.of(Ability.READ), requiredQueryId));
```

Comment:
Ability.READ.AS_SET

and make all API methods use Set<Ability> instead of EnumSet